### ISSUE TITLE
Potential use case for an Array of Arrays as keys 

### ISSUE THREAD

**Issue by SeanPercy** [graphql/dataloader#154]
I am currently working on a GraphQL application. Below are two of the types on my schema.
```
type Author {
  _id: ID!
  name: String!
  books: [Book!]!
}

type Book{
  _id: ID!
  title: String!
  books: [Author!]!
}

```
Most of my Book enties are only linked to one Author but since some of them are written by two or more Authors and I don't distinguish between main Author and co-Author(s), I ended up having an Array of Author._ids instead of just having one Author._id per Book. 

When I do a query as the following one:
```
query {
  allAuthors {
   books {
     title
    }
  }
}
```
I am going to fetch several Authors multiple times. Therefore I replaced the original `getByIds` function with a `DataLoader`, which essentially wraps the former one .
```
Book: {
  authors: async(book, _, context) => {
    return await context.dataloaders.authorsLoader.load(book.authors);
    //return await context.models.mongo.Author.getByIds(book.authors, context);
},
```

Here you see the my AuthorLoader.js file.
```
export const AuthorsLoader = (Authors) => {
  return new DataLoader(
    keys => batchAuthors(Authors, keys),
    {cacheKeyFn: key => key.toString()},
  );
};

function batchAuthors(Authors, keys) {
  return keys.map( book => {
    return Authors.findOne({_id: authorId});
    })
  });
}
```

The keys parameter is an Array of Arrays, with each sub-Array representing a Book instance. Each sub-Array contains at least one Author._id. So it looks kind of like this: `[ [auhtor1], [author1, author2], [author2, author1], [author1], [auhtor3], [auhtor3], [auhtor2], .... ]`
The code actually works in a sense that the query result is correct but then I realised that the keys being cached are sub-Arrays and therefore the order of Author._ids has to be taken care of before turning them into Strings or otherwise `[author1, author2]` and `[author2, author1]` will be two different keys.

Furthermore, can this approach considered to be a performative one?

**Comment by frederikhors** [comment]
@SeanPercy is this (https://github.com/facebook/dataloader/issues/155) the same problem?

**Comment by leebyron** [comment]
There's certainly nothing wrong with using an Array as a key, however you may want to use a `cacheKeyFn` to get proper caching behavior, especially since you may need to sort them to get a normalized form.
